integration-tests/smoke/ccip/fee_boosting_test.go

```diff
@@ -1,121 +1,303 @@
+	changeset.ConfirmExecWithSeqNrsForAll(t, e.Env, state, expectedSeqNumExec, startBlocks)
+}
+
+func willTriggerFeeBoosting(
+	t *testing.T,
+	msgSentEvent *onramp.OnRampCCIPMessageSent,
+	state changeset.CCIPOnChainState,
+	srcChain, destChain uint64) bool {
+	msg := convertToMessage(msgSentEvent.Message)
+	t.Log("\n=== Fee Boosting Analysis ===")
+	t.Logf("Src Chain: %d", msg.Header.SourceChainSelector)
+	t.Logf("Dest Chain: %d", msg.Header.DestChainSelector)
+
+	ep := ccipevm.NewGasEstimateProvider()
+	chainState, exists := state.Chains[srcChain]
+	require.True(t, exists)
+	feeQuoter := chainState.FeeQuoter
+
+	premium, err := feeQuoter.GetPremiumMultiplierWeiPerEth(&bind.CallOpts{Context: context.Background()}, chainState.Weth9.Address())
+	require.NoError(t, err)
+	t.Logf("Premium: %d", premium)
+
+	// Get LINK price
+	linkPrice, err := feeQuoter.GetTokenPrice(&bind.CallOpts{Context: context.Background()}, chainState.LinkToken.Address())
+	require.NoError(t, err)
+	t.Logf("LINK Price: %s", linkPrice.Value.String())
+	t.Logf("Juels in message: %s", msg.FeeValueJuels.String())
+
+	// Calculate fee in USD token
+	fee := new(big.Int).Div(
+		new(big.Int).Mul(linkPrice.Value, msg.FeeValueJuels.Int),
+		new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil),
+	)
+	t.Logf("Fee paid (in USD token): %s", fee.String())
+
+	// Calculate message gas
+	messageGas := new(big.Int).SetUint64(ep.CalculateMessageMaxGas(msg))
+	t.Logf("Estimated message gas: %s", messageGas.String())
+
+	// Get token and gas prices
+	nativeTokenAddress := chainState.Weth9.Address()
+	tokenAndGasPrice, err := feeQuoter.GetTokenAndGasPrices(&bind.CallOpts{Context: context.Background()}, nativeTokenAddress, destChain)
+	require.NoError(t, err)
+	t.Logf("Raw gas price (uint224): %s for chain: %d", tokenAndGasPrice.GasPriceValue.String(), destChain)
+
+	// Extract uint112 gas price
+	gasPrice, err := convertGasPriceToUint112(tokenAndGasPrice.GasPriceValue)
+	require.NoError(t, err)
+	t.Logf("Extracted gas price (uint112): %s", gasPrice.String())
+	t.Logf("Native token price: %s", tokenAndGasPrice.TokenPrice.String())
+
+	// Calculate total execution cost
+	execCost := new(big.Int).Mul(messageGas, gasPrice)
+	t.Logf("Total execution cost: %s", execCost.String())
+
+	// Check if fee boosting will trigger
+	willBoost := execCost.Cmp(fee) > 0
+	t.Logf("\nWill fee boosting trigger? %v", willBoost)
+	t.Logf("Execution cost / Fee ratio: %.2f",
+		new(big.Float).Quo(
+			new(big.Float).SetInt(execCost),
+			new(big.Float).SetInt(fee),
+		),
+	)
+
+	return execCost.Cmp(fee) > 0
+}
+
+func convertGasPriceToUint112(gasPrice *big.Int) (*big.Int, error) {
+	// Create a mask for uint112 (112 bits of 1s)
+	mask := new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 112), big.NewInt(1))
+
+	// Extract the lower 112 bits using AND operation
+	result := new(big.Int).And(gasPrice, mask)
+
+	return result, nil
+}
+
+func convertToMessage(msg onramp.InternalEVM2AnyRampMessage) cciptypes.Message {
```

Comment:
That's interesting, I thought we had code that does this already.

Reply:
I looked for one, but was not able to find it. It seems like something we already have...

Reply:
I think this code is folded somewhere in the ccipevm.Messagehasher or similar